engine/pkg/rpcerror/normalize.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcerror
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	pb "github.com/pingcap/tiflow/engine/enginepb"
+)
+
+// normalizeOpts stores information used to generate a Prototype.
+// Prototype cannot be elegantly used here, because its type parameter
+// would make it difficult to define NormalizeOpt.
+type normalizeOpts struct {
+	name string
+	msg  string
+}
+
+// NormalizeOpt is an option to Normalize.
+type NormalizeOpt func(opts *normalizeOpts)
+
+// WithMessage provides a custom error message.
+func WithMessage(msg string) NormalizeOpt {
+	return func(opts *normalizeOpts) {
+		opts.msg = msg
+	}
+}
+
+// WithName overrides the default error name, which
+// is the type name of the type argument to Normalize.
+func WithName(name string) NormalizeOpt {
+	return func(opts *normalizeOpts) {
+		opts.name = name
+	}
+}
+
+var prototypeRegistry = &sync.Map{}
+
+// Prototype is a generator for all instances
```

Comment:
IMHO,  a structual error called `normalizeError` has four members: `msg`, `name`, `isRetryable` and `grpcStatusCode`. 
Do we really need a `prototype` which has a generic param and `msg`, `name`?
Why not some methods for `normalizeError` for easy understanding.

Reply:
How do you distinguish an `error factory` and an instance of an error?

Reply:
Same question as https://github.com/pingcap/tiflow/pull/5950/files#r905973563